lib/Sema/CSDiagnostics.cpp

```diff
@@ -4099,7 +4099,8 @@ bool SubscriptMisuseFailure::diagnoseAsError() {
 
 bool SubscriptMisuseFailure::diagnoseAsNote() {
   if (auto overload = getOverloadChoiceIfAvailable(getLocator())) {
-    emitDiagnosticAt(overload->choice.getDecl(), diag::found_candidate);
+    auto decl = overload->choice.getDecl();
+    emitDiagnosticAt(decl, diag::found_candidate_type, decl->getOverloadSignatureType());
```

Comment:
You'll want to use `getInterfaceType` instead of `getOverloadSignatureType` (here and below), `getOverloadSignatureType` is what we use to determine if two decls overload each other, but may not actually have a type that reflects the decl, e.g all VarDecls have an overload type of `Void`.

I think we'll probably also want to drop the curried self from the type for type members, we probably ought to add a member to ValueDecl that returns this, the implementation would just be this: https://github.com/swiftlang/swift/blob/a5c6156525297716adefa58277d58d9ea8f6e63d/lib/Sema/AssociatedTypeInference.cpp#L2050-L2055